File: hackerrank/data-structures/stacks/simple-text-editor/simple-text-editor.pl
#!/usr/bin/env perl

# HackerRank - Simple Text Editor
# https://www.hackerrank.com/challenges/simple-text-editor
# with an undo stack of inverse actions

use strict;
use warnings;
use feature qw(say);
use experimental qw(signatures);

package Editor {
    sub new($class) {
        my $self = {buffer => q{}, undos => []};
        return bless $self, $class;
    }

    sub append_text($self, $text) {
        my $count = length $text;
        $self->{buffer} .= $text;
        push @{$self->{undos}}, sub { substr($self->{buffer}, -$count) = q{} };
        return;
    }

    sub shrink_by($self, $count) {
        my $text = substr $self->{buffer}, -$count, $count, q{};
        push @{$self->{undos}}, sub { $self->{buffer} .= $text };
        return;
    }

    sub print_character($self, $index) {
        say substr $self->{buffer}, $index - 1, 1;
        return;
    }

    sub undo($self) {
        pop(@{$self->{undos}})->();
        return;
    }
}

sub run() {
    my $editor = Editor->new();

    my $query_count = <>;

    for (1..$query_count) {
        my @tokens = grep { $_ } split /\s+/msx, <>;
        my $opcode = $tokens[0];

        if ($opcode == 1) {
            $editor->append_text($tokens[1]);
        } elsif ($opcode == 2) {
            $editor->shrink_by($tokens[1]);
        } elsif ($opcode == 3) {
            $editor->print_character($tokens[1]);
        } elsif ($opcode == 4) {
            $editor->undo();
        }
    }

    return;
}

if (!caller) { run() }

File: hackerrank/data-structures/stacks/simple-text-editor/simple-text-editor.js
#!/usr/bin/env node

// HackerRank - Simple Text Editor
// https://www.hackerrank.com/challenges/simple-text-editor
// naive approach, storing full snapshots

'use strict';

class Editor {
    constructor() {
        this._buffers = [''];
    }

    append(text) {
        this._buffers.push(this._current() + text);
    }

    truncate_by(count) {
        this._buffers.push(this._current().slice(0, -count));
    }

    print(index) {
        console.log(this._current()[index - 1]);
    }

    undo() {
        --this._buffers.length;
    }

    _current() {
        return this._buffers[this._buffers.length - 1];
    }
}

function main(getln) {
    const editor = new Editor();

    for (let q = parseInt(getln()); q > 0; --q) {
        const tokens = getln().split(/\s+/u).filter(tok => tok);

        switch (parseInt(tokens[0])) {
        case 1:
            editor.append(tokens[1]);
            break;

        case 2:
            editor.truncate_by(parseInt(tokens[1]));
            break;

        case 3:
            editor.print(parseInt(tokens[1]));
            break;

        case 4:
            editor.undo();
            break;
        }
    }
}

function run() {
    const reads = [];
    process.stdin.resume();
    process.stdin.setEncoding('utf-8');
    process.stdin.on('data', data => reads.push(data));
    process.stdin.on('end', () => {
        const lines = reads.join('').split('\n');
        let index = 0;
        main(() => lines[index++]);
    });
}

if (require.main === module) {
    run();
}
